### ISSUE TITLE
[1.12] CreativeCore compatibility issue

### ISSUE THREAD

**Issue by CreativeMD** [RootsTeam/Embers#237]
Hey there,

I'm the author of CreativeCore and somebody reported this issue: https://github.com/CreativeMD/CreativeCore/issues/18. I'm modifying the `loadedTileEntityList` of the world, to fix some horrible lag spikes (https://www.youtube.com/watch?v=vTCJbp45pWM).

So `loadedTileEntityList` is not really an `ArrayList` anymore, but instead a `HashMap<ChunkPos, List<TileEntity>>`. Actually you don't have to care about all this ...

Currently you are using the `size()` and `get(int index)` methods, which will not work with my changes:

https://github.com/RootsTeam/Embers/blob/28695d9b808e3470e0e6947df71d641285f9677d/src/main/java/teamroots/embers/EventManager.java#L586
```java
List<TileEntity> list = Minecraft.getMinecraft().world.loadedTileEntityList;
GlStateManager.pushMatrix();
for (int i = 0; i < list.size(); i ++){
	TileEntitySpecialRenderer render = TileEntityRendererDispatcher.instance.getSpecialRenderer(list.get(i));
	if (render instanceof ITileEntitySpecialRendererLater){
		double x = Minecraft.getMinecraft().player.lastTickPosX + Minecraft.getMinecraft().getRenderPartialTicks()*(Minecraft.getMinecraft().player.posX-Minecraft.getMinecraft().player.lastTickPosX);
		double y = Minecraft.getMinecraft().player.lastTickPosY + Minecraft.getMinecraft().getRenderPartialTicks()*(Minecraft.getMinecraft().player.posY-Minecraft.getMinecraft().player.lastTickPosY);
		double z = Minecraft.getMinecraft().player.lastTickPosZ + Minecraft.getMinecraft().getRenderPartialTicks()*(Minecraft.getMinecraft().player.posZ-Minecraft.getMinecraft().player.lastTickPosZ);
		((ITileEntitySpecialRendererLater)render).renderLater(list.get(i), list.get(i).getPos().getX()-x, list.get(i).getPos().getY()-y, list.get(i).getPos().getZ()-z, Minecraft.getMinecraft().getRenderPartialTicks());
	}
}
```
Instead you should use the iterator ...

```java
for(Iterator<TileEntity> iterator = Minecraft.getMinecraft().world.loadedTileEntityList.iterator(); iterator.hasNext()){
	TileEntity te = iterator.next();
	TileEntitySpecialRenderer render = TileEntityRendererDispatcher.instance.getSpecialRenderer(te);
	if (render instanceof ITileEntitySpecialRendererLater){
		double x = Minecraft.getMinecraft().player.lastTickPosX + Minecraft.getMinecraft().getRenderPartialTicks()*(Minecraft.getMinecraft().player.posX-Minecraft.getMinecraft().player.lastTickPosX);
		double y = Minecraft.getMinecraft().player.lastTickPosY + Minecraft.getMinecraft().getRenderPartialTicks()*(Minecraft.getMinecraft().player.posY-Minecraft.getMinecraft().player.lastTickPosY);
		double z = Minecraft.getMinecraft().player.lastTickPosZ + Minecraft.getMinecraft().getRenderPartialTicks()*(Minecraft.getMinecraft().player.posZ-Minecraft.getMinecraft().player.lastTickPosZ);
		((ITileEntitySpecialRendererLater)render).renderLater(te, te.getPos().getX()-x, te.getPos().getY()-y, te.getPos().getZ()-z, Minecraft.getMinecraft().getRenderPartialTicks());
	}
}
```

Theoretically i could fix this issue from my end, but `get(int index)` is incredible bad for performance, basically it would cause horrible lag, so i rather cause a crash.

It would be very cool if you could implement this change.

In Regards
CreativeMD

**Comment by elucent** [comment]
I'll try it out, but the issue with an iterator is that it could leave me open to CMEs I otherwise wouldn't have to deal with. Particularly for a TileEntity list being handled on the client, when a packet could come at any time and delete a TE out of it and break the iteration.

Will toss this into the next Embers build whenever it happens but I'm on a little break from my mods ATM so it won't be right away.

**Comment by elucent** [comment]
Actually, what do you need the Map<ChunkPos, TileEntity> for? If it's what I think it is, I may have a solution.

**Comment by CreativeMD** [comment]
None of these solutions are thread safe, but they will not cause a crash if new tileentities will be added during the iteration, but they will most likely crash if a tileentity has been removed.
During every tick the world will delete all unloaded tileentities. The field `processingLoadedTiles` of the world indicates whether those list are currently modified or not.

In your case you don't have to take care of it, since both of it happens in the same thread, so it cannot happen simultaneously. So there is no threading issue or what so ever.

Have you watched the video? It shows a huge lag spike which occurs if you unload chunks with a lot of tileentities in it. The chunk will add all of his tileentities to a list, which the world will remove during the tick. This is extremely bad for performance, in my example it deleted 3000 out of a list of 20000, which takes extremely long. So i modified the system. The HashMap is sorted to chunk positions to make it faster to remove a specific tileentitiy. It also allows you to remove a complete chunk at once, which is a hell of a lot quicker. So basically this will fix the lag spike.

As i said earlier i could fix it from my end, but this would be extremely bad for performance so i rather make it crash.

**Comment by elucent** [comment]
If it's simply an unrelated performance boost, you ought to make it optional on your end and warn about it causing potential problems. If you do already, great.

But yeah, as I said, I can implement a fix. A for loop with an integer index _would_ be thread-safe in my case: if a TE gets deleted, it'll simply close the loop prematurely. For a frame, some tiles might not render, but that's a near trivial concern. I'll try out your way though and see if it causes issues.

**Comment by CreativeMD** [comment]
Ok would be cool :D

Although i have to say there are several ways an index iteration can crash:
* The index can be too large (if one or more entries are deleted)
* Multiple get(index) calls can have different results, which could also cause some crashes

**Comment by TehNut** [comment]
Would this be solved by https://github.com/SleeplessAcorn/Embers/commit/e9db16cacfbab734eb68815e003fcccc221a42f8?

**Comment by CreativeMD** [comment]
Yes of course, although the loop of loadedTileEntityList has to be changed.

**Comment by TehNut** [comment]
Ah, right. This was an automated cleanup and I got excited I may have accidentally fixed it

**Comment by kreezxil** [comment]
This is also happening 1.11.2-2.2.0 https://pastebin.com/3azYYsZV
